modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,252 @@
+    else if (strncmp(*argv, "fs=", 3) == 0)
+      p->fs = (char *)*argv + 3;
+    else if (strncmp(*argv, "overwrite=", 10) == 0)
+      p->overwrite = strtol(*argv + 10, NULL, 10);
+    else if (strncmp(*argv, "debug=", 6) == 0)
+      p->debug = strtol(*argv + 6, NULL, 10);
+  }
+}
+
+PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc,
+                                   const char **argv) {
+  int retval;
+  char *val, *mntdevice;
+  const void *user;
+  const struct passwd *pwd;
+  struct pam_params param = { .start_uid = 1000, .end_uid = 0, .fs = NULL };
+  struct if_dqblk ndqblk;
+  FILE *fp;
+  size_t match_size = 0;
+  const struct mntent *mnt;
+  const char *service;
+
+  if (pam_get_item(pamh, PAM_SERVICE, (const void **)&service) != PAM_SUCCESS)
+    service = "";
+
+  /* Get UID_MIN for default start_uid from login.defs */
+  val = pam_modutil_search_key(pamh, LOGIN_DEFS, "UID_MIN");
+
+  /* Should UID_MIN be undefined, use 1000 as safe starting UID
+   * to avoid setting a quota for root and system users if startuid= parameter
+   * is absent.
+   */
+  if (val) {
+    param.start_uid = strtol(val, NULL, 10);
+    free(val);
+  }
+
+  /* Parse values */
+  _pam_parse_params(argc, argv, &param);
+
+  /* Determine the user name so we can get the home directory */
+  retval = pam_get_item(pamh, PAM_USER, &user);
+  if (retval != PAM_SUCCESS || user == NULL || *(const char *)user == '\0') {
+    pam_syslog(pamh, LOG_NOTICE, "user unknown");
+    return PAM_USER_UNKNOWN;
+  }
+
+  /* Get the password entry */
+  pwd = pam_modutil_getpwnam(pamh, user);
+  if (pwd == NULL) {
+    return PAM_CRED_INSUFFICIENT;
+  }
+
+  /* Check if we should not set quotas for user */
+  if ((pwd->pw_uid < param.start_uid) ||
+      ((param.end_uid >= param.start_uid) && (param.start_uid != 0) &&
+       (pwd->pw_uid > param.end_uid)))
+    return PAM_SUCCESS;
+
+  /* Remove the unnecessary '/' from the end of fs parameter */
+  if (param.fs != NULL) {
+    size_t len = strlen(param.fs);
+    if (len > 1)
+      if (param.fs[len - 1] == '/')
+        param.fs[len - 1] = '\0';
+  }
+
+  /* Find out what device the filesystem is hosted on */
+  if ((fp = setmntent("/etc/mtab", "r")) == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Unable to open /etc/mtab");
+    return PAM_PERM_DENIED;
+  }
+
+  *mntdevice = '\0';
+  while ((mnt = getmntent(fp)) != NULL) {
+    if (param.fs == NULL) {
+      size_t mnt_len = strlen(mnt->mnt_dir);
+      /* If fs is not specified use filesystem with homedir as default
+       * Checking the mnt_len-th character in pwd->pw_dir is safe because of the
+       * strncmp(2) check before (whose success implies strlen(pwd->pw_dir) >=
+       * mntlen)
+       */
+      if ((strncmp(pwd->pw_dir, mnt->mnt_dir, mnt_len) == 0) &&
+          (mnt_len > match_size) &&
+          (pwd->pw_dir[mnt_len] == '\0' || pwd->pw_dir[mnt_len] == '/')) {
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
```

Comment:
Please ```free(mntdevice)``` first.

Reply:
Pardon my ignorance, but if mntdevice has been initialized to NULL and not used before, why would I need to free() it before using strdup();

Reply:
What if it was assigned before?
```free(NULL)``` is a no-op.

Reply:
I see my comprehension problem here. I totally blanked out that we are in a loop here, so of course mntdevice could have been strdup'ed before.
Again, thank you for being patient with me. My normal coding environment is Perl and Python, my C-foo is a bit rusty after 15 years not using it.